host_applications/linux/apps/raspicam/gl_scenes/shader.c

```diff
@@ -0,0 +1,287 @@
+
+    //~ offset += 0.05;
+    GLCHK(glUseProgram(shader.program));
+    GLCHK(glEnableVertexAttribArray(shader.attribute_locations[0]));
+    GLfloat varray[] = {
+        -1.0f, -1.0f,
+        1.0f,  1.0f,
+        1.0f, -1.0f,
+
+        -1.0f,  1.0f,
+        1.0f,  1.0f,
+        -1.0f, -1.0f,
+    };
+    GLCHK(glVertexAttribPointer(shader.attribute_locations[0], 2, GL_FLOAT, GL_FALSE, 0, varray));
+    
+    shader_update_shader(state);
+
+    //GLCHK(glUniform1f(shader.uniform_locations[1], offset));
+    GLCHK(glDrawArrays(GL_TRIANGLES, 0, 6));
+
+    GLCHK(glDisableVertexAttribArray(shader.attribute_locations[0]));
+    GLCHK(glUseProgram(0));
+    return 0;
+}
+
+int shader_open(RASPITEX_STATE *state)
+{
+   const GLchar *defaultVertexShader = 
+    "attribute vec2 vertex;\n"
+    "varying vec2 texcoord;"
+    "void main(void) {\n"
+    "   texcoord = 0.5 * (vertex + 1.0);\n"
+    "   gl_Position = vec4(vertex, 0.0, 1.0);\n"
+    "}\n";
+
+   const GLchar *defaultFragmentShader =  
+      "#extension GL_OES_EGL_image_external : require\n"
+      "varying vec2 texcoord;"
+      "uniform samplerExternalOES tex;"
+      "void main() \n"
+      "{ \n"
+      " gl_FragColor = texture2D(tex, texcoord); \n"
+      "} \n";
+
+   FILE * pfFile=NULL;
+   FILE * pvFile=NULL;
+   long lSize;
+   int result;
+
+   pfFile = fopen (state->fragment_shader_filename,"rb");
+   if (pfFile==NULL) {
+      printf("No or invalid fragment file provided, used the default one\n");
+      shader.fragment_source=defaultFragmentShader;
+   } else {
+
+   // obtain file size:
+   fseek (pfFile , 0 , SEEK_END);
+   lSize = ftell (pfFile);
+   rewind (pfFile);
+
+   // allocate memory to contain the whole file:
+   shader.fragment_source = (GLchar*) malloc (sizeof(GLchar)*lSize);
+   if (shader.fragment_source == NULL) {fputs ("Memory error",stderr); exit (2);}
+
+   // copy the file into the buffer:
+   result = fread (shader.fragment_source,1,lSize,pfFile);
+   if (result != lSize) {fputs ("Reading error",stderr); exit (3);}
+   
+   printf("fragment shader file : %s\n", state->fragment_shader_filename);
+
+   /* the whole file is now loaded in the memory buffer. */
+   }
+
+   pvFile = fopen (state->vertex_shader_filename,"rb");
+   if (pvFile==NULL) {
+      printf("No or invalid vertex file provided, used the default one\n");
+      shader.vertex_source=defaultVertexShader;
+   } else {
+
+   // obtain file size:
+   fseek (pvFile , 0 , SEEK_END);
+   lSize = ftell (pvFile);
+   rewind (pvFile);
+
+   // allocate memory to contain the whole file:
+   shader.vertex_source = (GLchar*) malloc (sizeof(GLchar)*lSize);
+   if (shader.vertex_source == NULL) {fputs ("Memory error",stderr); exit (2);}
+
+   // copy the file into the buffer:
+   result = fread (shader.vertex_source,1,lSize,pvFile);
+   if (result != lSize) {fputs ("Reading error",stderr); exit (3);}
+   
+   printf("vertex shader file : %s\n", state->vertex_shader_filename);
+
```

Comment:
It would be cleaner to return an erorr from this function instead of just exiting. RaspiStill might want to do explicit cleanup